robyn/templating.py

```diff
@@ -1,22 +1,84 @@
+from typing import Callable, List
 from abc import ABC, abstractmethod
 
 from jinja2 import Environment, FileSystemLoader
 
-from robyn import status_codes
+from robyn import status_codes, Robyn
+from robyn.router import Route
+# from robyn import robyn
 
 from .robyn import Headers, Response
 
 
+def get_param_filled_url(url: str, kwdict: dict | None = None) -> str:
+    """fill the :params in the url
+
+    Args:
+        url (str): typically comes from the route
+        kwdict (dict): the **kwargs as a dict
+
+    Returns:
+        str: _description_modified url (if there are elements in kwdict, otherwise unchanged)
+    """
+    if kwdict is not None:
+        for k, v in zip(kwdict.keys(), kwdict.values()):
+            url = url.replace(f":{k}", f"{v}")
+
+    return url
+
+
 class TemplateInterface(ABC):
     def __init__(self): ...
 
     @abstractmethod
     def render_template(self, *args, **kwargs) -> Response: ...
 
+    @abstractmethod
+    def set_robyn(self, robyn: Robyn) -> None: ...
+
+    @abstractmethod
+    def get_function_url(self, function_name: str, route_type: str = "GET", **kwargs) -> str: ...
+
 
 class JinjaTemplate(TemplateInterface):
-    def __init__(self, directory, encoding="utf-8", followlinks=False):
-        self.env = Environment(loader=FileSystemLoader(searchpath=directory, encoding=encoding, followlinks=followlinks))
+    def __init__(self, directory, encoding="utf-8", followlinks=False) -> None:
+        self.env: Environment = Environment(loader=FileSystemLoader(searchpath=directory, encoding=encoding, followlinks=followlinks))
+        self.add_function_to_globals("get_function_url", self.get_function_url)
+        self.robyn: Robyn | None = None
+
+    def add_function_to_globals(self, name: str, func: Callable):
+        """
+        Add a global function to a Jinja environment.
+        """
+        self.env.globals[name] = func
+
+    def set_robyn(self, robyn: Robyn) -> None:
+        """
+        The get_function_url needs to have access to the list of routes stored in the apps Robyn object
+
+        Args:
+            robyn (Robyn): The top instance of the Robyn class for this app.
+        """
```

Comment:
I am sorry if I am being a bit daft but why is this function needed?

Reply:
When we are rendering a template that includes a call into the TemplateInterface `get_function_url(...)` that function needs to get the live and complete list of routes. That means we can't have the TemplateInterface keep a copy of the routes that existed when it was created. Instead, when the call is made we check the live list of routes. The only place we can get that from is the main app Robyn instance.

I couldn't see a way to get old of the Robyn instance automagically. So it was either a set method or the constructor.

I decided to go for a set method as the constructor for JiujaTemplate already has 3 arguments plus self (I figured it will impact the existing code less and be easier to add additions templating packages)

If projects don't use `get_function_url(...)` then they don't need to call set_robyn.